src/aiortc/contrib/media.py

```diff
@@ -560,3 +560,90 @@ async def __run_track(self, track: MediaStreamTrack) -> None:
         self.__log_debug("Stop reading source %s", id(track))
         del self.__proxies[track]
         del self.__tasks[track]
+
+class NativeRelayStreamTrack(MediaStreamNativeTrack):
```

Comment:
This is a quick-and-dirty way to get things to work -- I just copied `RelayStreamTrack` and `MediaRelay` and made `Native*` versions of them that use `MediaStreamNativeTrack`.

It'd probably be better to refactor out base classes with the logic, then the subclasses can switch out the type.

Reply:
Yes, use `Union` structure to receive `MediaStreamNativeTrack` and `MediaStreamTrack`.

```
class RelayStreamTrack(Union[MediaStreamTrack, MediaStreamNativeTrack]):
    def __init__(self, relay, source: Union[MediaStreamTrack, MediaStreamNativeTrack]) -> None:
        super().__init__()
        self.kind = source.kind
        self._relay = relay
        self._queue: asyncio.Queue[Optional[Frame]] = asyncio.Queue()
        self._source: Union[Optional[MediaStreamTrack], Optional[MediaStreamNativeTrack]] = source

    async def recv(self):
        if self.readyState != "live":
            raise MediaStreamError

        self._relay._start(self)
        frame = await self._queue.get()
        if frame is None:
            self.stop()
            raise MediaStreamError
        return frame

    def stop(self):
        super().stop()
        if self._relay is not None:
            self._relay._stop(self)
            self._relay = None
            self._source = None


class MediaRelay:
    """
    A media source that relays one or more tracks to multiple consumers.

    This is especially useful for live tracks such as webcams or media received
    over the network.
    """

    def __init__(self) -> None:
        self.__proxies: Dict[Union[MediaStreamTrack, MediaStreamNativeTrack], Set[RelayStreamTrack]] = {}
        self.__tasks: Dict[Union[MediaStreamTrack, MediaStreamNativeTrack], asyncio.Future[None]] = {}

    def subscribe(
        self, 
        track: Union[MediaStreamTrack, MediaStreamNativeTrack]
    ) -> Union[MediaStreamTrack, MediaStreamNativeTrack]:
        """
        Create a proxy around the given `track` for a new consumer.
        """
        proxy = RelayStreamTrack(self, track)
        self.__log_debug("Create proxy %s for source %s", id(proxy), id(track))
        if track not in self.__proxies:
            self.__proxies[track] = set()
        return proxy

    def _start(self, proxy: RelayStreamTrack) -> None:
        track = proxy._source
        if track is not None and track in self.__proxies:
            # register proxy
            if proxy not in self.__proxies[track]:
                self.__log_debug("Start proxy %s", id(proxy))
                self.__proxies[track].add(proxy)

            # start worker
            if track not in self.__tasks:
                self.__tasks[track] = asyncio.ensure_future(self.__run_track(track))

    def _stop(self, proxy: RelayStreamTrack) -> None:
        track = proxy._source
        if track is not None and track in self.__proxies:
            # unregister proxy
            self.__log_debug("Stop proxy %s", id(proxy))
            self.__proxies[track].discard(proxy)

    def __log_debug(self, msg: str, *args) -> None:
        logger.debug(f"MediaRelay(%s) {msg}", id(self), *args)

    async def __run_track(self, track: Union[MediaStreamTrack, MediaStreamNativeTrack]) -> None:
        self.__log_debug("Start reading source %s" % id(track))

        while True:
            try:
                frame = await track.recv()
            except MediaStreamError:
                frame = None
            for proxy in self.__proxies[track]:
                proxy._queue.put_nowait(frame)
            if frame is None:
                break

        self.__log_debug("Stop reading source %s", id(track))
        del self.__proxies[track]
        del self.__tasks[track]
```

Test with this modification.

Reply:
I don't think you can subclass a Union type:

```python
class RelayStreamTrack(Union[MediaStreamTrack, MediaStreamNativeTrack]):
```

But maybe could do something similar with python [generics](https://stackoverflow.com/questions/6725868/generics-templates-in-python/59589910)?

Reply:
I was at least able to apply your changes to get rid of `NativeMediaRelay`. Those are pushed to this branch now. Will need to think more about how to get rid of `RelayStreamNativeTrack` since it's subclassed from `MediaStreamNativeTrack`

Reply:
Great! I'll incorporate this changes in my branch. ok? I'll merge these modification.

Reply:
Awesome!!